src/mock_server.jl

```diff
@@ -0,0 +1,276 @@
+using CloudBase: CloudCredentials, AWSCredentials, AbstractStore, AWS
+using JSON3, HTTP, Sockets, Base64
+using RustyObjectStore: SnowflakeConfig, ClientOptions
+
+export SFGatewayMock, start, with
+
+struct SFConfig
+    account::String
+    database::String
+    schema::String
+    stage::String
+end
+
+mutable struct SFGatewayMock
+    credentials::CloudCredentials
+    store::AbstractStore
+    opts::ClientOptions
+    config::SFConfig
+    encrypted::Bool
+    keys_lock::ReentrantLock
+    next_key_id::Int
+    keys::Dict{String, String}
+end
+
+function SFGatewayMock(credentials::CloudCredentials, store::AbstractStore, encrypted::Bool; opts=ClientOptions())
+    stage_name = "teststage" * string(rand(UInt64), base=16)
+    config = SFConfig(
+        "testaccount",
+        "testdatabase",
+        "testschema",
+        stage_name
+    )
+    SFGatewayMock(
+        credentials,
+        store,
+        opts,
+        config,
+        encrypted,
+        ReentrantLock(),
+        1,
+        Dict{String, String}()
+    )
+end
+
+function authorized(request::HTTP.Request)
+    return HTTP.header(request, "Authorization") == "Snowflake Token=\"dummy-token\""
+end
+
+function unauthorized_response()
+    return HTTP.Response(401, "Invalid token")
+end
+
+function error_json_response(msg::String; code::String="1234")
+    response_data = Dict(
+        "data" => Dict(
+            "queryId" => "dummy-query-id"
+        ),
+        "code" => code,
+        "message" => msg,
+        "success" => false
+    )
+    return HTTP.Response(200, JSON3.write(response_data))
+end
+
+function construct_stage_info(credentials::AWSCredentials, store::AWS.Bucket, path::String, encrypted::Bool)
+    m = match(r"(https?://.*?)/", store.baseurl)
+    @assert !isnothing(m)
+    test_endpoint = m.captures[1]
+    @show test_endpoint
+
+    Dict(
+        "locationType" => "S3",
+        "location" => joinpath(store.name, "xox50000-s", path),
+        "path" => path,
+        "region" => "us-east-1",
+        "storageAccount" => nothing,
+        "isClientSideEncrypted" => encrypted,
+        "ciphers" => encrypted ? "AES_CBC" : nothing,
+        "creds" => Dict(
+            "AWS_KEY_ID" => credentials.access_key_id,
+            "AWS_SECRET_KEY" => credentials.secret_access_key,
+            "AWS_TOKEN" => credentials.session_token
+        ),
+        "useS3RegionalUrl" => false,
+        "endPoint" => "us-east-1.s3.amazonaws.com",
+        "testEndpoint" => test_endpoint
+    )
+end
+
+function next_id_and_key(gw::SFGatewayMock)
+    @lock gw.keys_lock begin
+        key_id = gw.next_key_id
+        gw.next_key_id += 1
+        key = base64encode(rand(UInt8, 16))
+        push!(gw.keys, string(key_id) => key)
+        return key_id, key
+    end
+end
+
+function find_key_by_id(gw::SFGatewayMock, id::String)
+    @lock gw.keys_lock begin
+        return get(gw.keys, id, nothing)
+    end
+end
+
+function start(gw::SFGatewayMock)
```

Comment:
Returns a mocked SnowflakeConfig and an alive server.
Using the config, we can ask for the credentials and keys that we need to e.g.,  write pages to minio